azure-spring-boot/src/main/java/com/microsoft/azure/keyvault/spring/secrets/KeyVaultSecretTemplate.java

```diff
@@ -0,0 +1,94 @@
+import com.microsoft.azure.keyvault.models.SecretItem;
+import com.microsoft.azure.keyvault.spring.AbstractKeyVaultTemplate;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * Default implementation of {@link KeyVaultSecretOperation}
+ *
+ */
+public class KeyVaultSecretTemplate extends AbstractKeyVaultTemplate implements KeyVaultSecretOperation {
+    private final Map<String, Set<String>> secretNamesByKeyVault = new ConcurrentHashMap<>();
+    private final Map<Pair<String, String>, String> secretsByKeyVault = new ConcurrentHashMap<>();
+
+    private long lastUpdatedTime = System.currentTimeMillis();
+
+    public KeyVaultSecretTemplate(String clientId, String clientSecret) {
+        super(clientId, clientSecret);
+    }
+
+    public KeyVaultSecretTemplate(String clientId, String clientSecret, long refreshIntervalMS) {
+        super(clientId, clientSecret, refreshIntervalMS);
+    }
+
+    @Override
+    public String getSecret(String keyVaultName, String secretName) {
+        final Pair<String, String> keyVaultAndSecret = Pair.of(keyVaultName, secretName);
+        if (useCache) {
+            clearCacheAfterInterval();
+            this.secretsByKeyVault.computeIfAbsent(keyVaultAndSecret, this::fetchSecret);
+        } else {
+            this.secretsByKeyVault.put(keyVaultAndSecret, fetchSecret(keyVaultAndSecret));
+        }
+
+        return this.secretsByKeyVault.get(keyVaultAndSecret);
+    }
+
+    @Override
+    public List<String> listSecrets(String keyVaultName) {
+        if (useCache) {
+            clearCacheAfterInterval();
+            this.secretNamesByKeyVault.computeIfAbsent(keyVaultName, this::fetchSecretNames);
+        } else {
+            this.secretNamesByKeyVault.put(keyVaultName, fetchSecretNames(keyVaultName));
+        }
+
+        return new ArrayList<>(this.secretNamesByKeyVault.get(keyVaultName));
+    }
+
+    private Set<String> fetchSecretNames(String keyVaultName) {
+        final String keyVaultUrl = buildKeyVaultUrl(keyVaultName);
+        final PagedList<SecretItem> secretItems =
+                this.keyVaultClientCreator.apply(keyVaultName).listSecrets(keyVaultUrl);
+
+        return Collections.unmodifiableSet(
+                secretItems.stream().map(SecretItem::id).map(s -> s.substring(s.lastIndexOf('/') + 1))
+                        .collect(Collectors.toSet()));
+    }
+
+    private String fetchSecret(Pair<String, String> keyVaultAndSecret) {
+        final String keyVaultName = keyVaultAndSecret.getLeft();
+
+        final SecretBundle bundle = this.keyVaultClientCreator.apply(keyVaultName)
+                .getSecret(buildKeyVaultUrl(keyVaultName), keyVaultAndSecret.getRight());
+
+        return bundle == null ? null : bundle.value();
+    }
+
+    private void clearCacheAfterInterval() {
+        if (refreshIntervalMS <= 0) {
```

Comment:
If we want to refresh periodically, I suggest to use separate scheduled thread pool to do this. This will improve latency when user really want to use such value.

Reply:
Wonder is it too heavy to introduce thread pool here.

Reply:
Even though we refresh periodically, but the secrets still have to be queried, there is no existing way to check updates only.